include/boost/graph/link_cut_trees.hpp

```diff
@@ -0,0 +1,280 @@
+//
+//=======================================================================
+// Copyright 2019
+// Author: Yi Ji
+//
+// Distributed under the Boost Software License, Version 1.0. (See
+// accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//=======================================================================
+//
+#ifndef BOOST_LINK_CUT_TREES_HPP
+#define BOOST_LINK_CUT_TREES_HPP
+
+#include <boost/graph/properties.hpp>
+#include <boost/unordered_map.hpp>
+
+namespace boost
+{
+    template<class ElementParentMap, class ElementChildMap = ElementParentMap>
+    class link_cut_trees
+    {
+    public:
+        inline link_cut_trees(ElementParentMap p, ElementChildMap l, ElementChildMap r) : parent(p), left(l), right(r) {}
+        
+        inline link_cut_trees(const link_cut_trees<ElementParentMap, ElementChildMap> &c)
+            : parent(c.parent), left(c.left), right(c.right) {}
+
+        template <class Element>
+        inline void make_tree(Element x)
+        {
+            make_path(x);
+        }
+
+        template <class Element>
+        inline Element find_root(Element x)
+        {
+            return find_tail(expose(x));
+        }
+
+        template <class Element>
+        inline void link(Element x, Element y) // Element x must be a tree root
+        {
+            Element r = expose(x);
+            r = join(r, r, expose(y));
+            put_successor(r, r);
+        }
+
+        template <class Element>
+        inline void cut(Element x)
+        {
+            expose(x);
+            std::pair<Element, Element> uv = split(x);
+            put_successor(x, x);
+            put_successor(uv.second, uv.second);
+        }
+
+        template <class Element>
+        inline Element lowest_common_ancestor(Element x, Element y) // Elements x and y must have same root
+        {
```

Comment:
Again, something like `BOOST_ASSERT(find_root(x) == find_root(y));`

Reply:
Done adding it.